### PR TITLE
auto save model in lightning

### PR DESCRIPTION
**Pull request by dberenbaum**
This PR auto logs models with `dvclive.lightning.DVCLiveLogger(log_model=True)`:

```
trainer = pl.Trainer(logger=[DVCLiveLogger(log_model=True)])
```

`log_model` follows the conventions in [mlflow](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.loggers.mlflow.html#module-lightning.pytorch.loggers.mlflow) and [wandb](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.loggers.wandb.html):
* `False` saves no models (this is the default).
* `True` saves all model checkpoints at the end of training. 
* `"all"` saves all model checkpoints whenever a model checkpoint is saved.

If `log_model` is `True` or `"all"`, dvclive caches the entire checkpoints folder. 

Dvclive will also add a model artifact named "best" at the end of training that references the best model checkpoint. (edit: this resembles the [best alias](https://docs.wandb.ai/guides/integrations/lightning#model-checkpointing) in wandb)

Edit: example `dvclive/dvc.yaml` output:

```yaml
artifacts:
  best:
    path: ../DvcLiveLogger/dvclive_run/checkpoints/epoch=1-step=4-v1.ckpt
    type: model
```

To support this, `log_artifact` was also changed:
* Artifacts will only be added to `dvc.yaml:artifacts` if some metadata is provided (type, name, desc, labels, or meta). This is a breaking change, but I can't see how anyone is making use of this without any metadata since it won't be used by the model registry.
* Added `cache` kwarg to `log_artifact` (defaults to `True`) so that it's possible to add the artifact metadata without caching the object.

Related:
* #504 
* #551 
* #572 
* #586 

To do:
- [x] Tests
- [ ] Docs (https://github.com/iterative/dvc.org/pull/4714)
- [ ] VS Code snippets (cc @mattseddon)
- [ ] Studio snippets (cc @yathomasi)

### CODE REVIEW

**Comment by dberenbaum** [comment]
Overall, I felt opening a PR with the desired behavior would be more effective than explaining and discussing in an issue. I hope this will help resolve some of the rough edges around saving models and that we can work through the other framework callbacks to implement similar functionality that works with the existing framework conventions and resembles mlflow, wandb, etc.

**Comment by dberenbaum** [comment]
One thing to note: lightning will not overwrite existing files or clean up between runs. Instead, it will append a version number, so if you run the same code repeatedly, you will end up with a directory that tracks your entire history of model checkpoints instead of only the latest run:

```
$ tree DvcLiveLogger
DvcLiveLogger
└── dvclive_run
    ├── checkpoints
    │   ├── epoch=0-step=2-v1.ckpt
    │   ├── epoch=0-step=2-v2.ckpt
    │   ├── epoch=0-step=2.ckpt
    │   ├── epoch=1-step=32-v1.ckpt
    │   ├── epoch=1-step=32-v2.ckpt
    │   ├── epoch=1-step=32.ckpt
    │   ├── epoch=1-step=4.ckpt
    │   ├── epoch=2-step=6.ckpt
    │   ├── epoch=4-step=10-v1.ckpt
    │   ├── epoch=4-step=10-v2.ckpt
    │   ├── epoch=4-step=10-v3.ckpt
    │   ├── epoch=4-step=10-v4.ckpt
    │   ├── epoch=4-step=10-v5.ckpt
    │   ├── epoch=4-step=10-v6.ckpt
    │   └── epoch=4-step=10.ckpt
    └── checkpoints.dvc
```

If you are running a pipeline, this is probably fine since you can control if you want to delete that directory each time. We might also want to consider dropping the existing checkpoints directory in the dvclive callback if `resume=False`.

**Comment by daavoo** [comment]
> log_model follows the conventions in [mlflow](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.loggers.mlflow.html#module-lightning.pytorch.loggers.mlflow) and [wandb](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.loggers.wandb.html):

`True` for MLFLow appears to be different:

```
log_model follows the conventions in [mlflow](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.loggers.mlflow.html#module-lightning.pytorch.loggers.mlflow) and [wandb](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.loggers.wandb.html):
```

**Comment by dberenbaum** [comment]
@daavoo I can add the behavior for `save_top_k == -1` to log every checkpoint. I figured it's a minor detail and a quick google search didn't turn up any justification for why they do it, but wandb does it also, so we can do it for consistency.

**Comment by dberenbaum** [comment]
I think this gets us to a good place with logging models in lightning. In fact, comparing to other trackers, it feels a bit easier to manage the models this way in dvc. On a different machine, you can pull the lightning checkpoints dir and keep using lightning methods to load those checkpoints. With other trackers, once you are on a different machine, the only way to load models is using the experiment tracker's api.

File: src/dvclive/lightning.py [hunk 1]
```diff
@@ -119,6 +123,23 @@ def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None):
...
+        ):
+            self.experiment.log_artifact(checkpoint_callback.dirpath)
+
     @rank_zero_only
     def finalize(self, status: str) -> None:
+        checkpoint_callback = self._checkpoint_callback
+        # Save model checkpoints.
+        if self._log_model is True:
+            self.experiment.log_artifact(checkpoint_callback.dirpath)
+        # Log best model.
```
Comment (daavoo) [inline src/dvclive/lightning.py:139]: WDYT of creating a copy in "dvclive" folder (or in the checkpoints folder itself), at least for the best?

It seems that we would be changing the path of the registered model between experiments in the current behavior

**Comment by daavoo** [comment]
> If you are running a pipeline, this is probably fine since you can control if you want to delete that directory each time. We might also want to consider dropping the existing checkpoints directory in the dvclive callback if resume=False.

Didn't look in details, but seems like the other loggers use [_scan_checkpoints ](https://github.com/Lightning-AI/lightning/blob/017262e5e0c65215e9e75121d155d7a07cd9e7bf/src/lightning/pytorch/loggers/utilities.py#L32) to only track the ones related to the current experiment

File: src/dvclive/lightning.py (hunk 1 above)
Comment (dberenbaum) [reply src/dvclive/lightning.py:139]: Yes, which I guess is also what other trackers do AFAICT? Do you think the path matters? Maybe it makes it easier to `dvc get` later, although we could make that work by the artifact name. No strong opinion from me.

**Comment by dberenbaum** [comment]
> Didn't look in details, but seems like the other loggers use [_scan_checkpoints ](https://github.com/Lightning-AI/lightning/blob/017262e5e0c65215e9e75121d155d7a07cd9e7bf/src/lightning/pytorch/loggers/utilities.py#L32) to only track the ones related to the current experiment

Great idea. I'll look into it.

**Comment by dberenbaum** [comment]
Lightning [warns](https://github.com/Lightning-AI/lightning/blob/017262e5e0c65215e9e75121d155d7a07cd9e7bf/src/lightning/pytorch/callbacks/model_checkpoint.py#L620) if the directory is not empty:

```console
UserWarning: Checkpoint directory /Users/dave/Code/lstm_seq2seq/model exists and is not empty.
  rank_zero_warn(f"Checkpoint directory {dirpath} exists and is not empty.")
```

`_scan_checkpoints` only returns the latest checkpoint even if you use something like `ModelCheckpoint(save_top_k=-1)`, so I had to save the checkpoint from each scan and then drop the rest of the files in the directory (unless `resume=True`).

Removing the files only happens after the checkpoint is saved, so sometimes the first checkpoint will still get a version number like:
```console
$ tree model
model
├── epoch=0-step=2-v1.ckpt # saved this checkpoint before previous one was dropped
├── epoch=1-step=4.ckpt
├── epoch=2-step=6.ckpt
├── epoch=3-step=8.ckpt
└── epoch=4-step=10.ckpt
```

Overall, this works and probably meets most user's expectations, so I think we should keep it, but I don't feel strongly that it outweighs the added complexity or potential surprise that dvclive is deleting model checkpoints.

**Comment by dberenbaum** [comment]
ping @daavoo

File: src/dvclive/lightning.py [hunk 2]
```diff
@@ -131,6 +141,42 @@ def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None):
...
+        if self._log_model == "all" or (
+            self._log_model is True and checkpoint_callback.save_top_k == -1
+        ):
+            self._save_checkpoints(checkpoint_callback)
+
     @rank_zero_only
     def finalize(self, status: str) -> None:
+        # Log best model.
+        if self._checkpoint_callback:
+            self._scan_checkpoints(self._checkpoint_callback)
```
Comment (daavoo) [inline src/dvclive/lightning.py:157]:
```suggestion
```

**Review by daavoo — APPROVED** [review]
On a high level, the code and description make sense to me.

I have not actually tried in a project the different options, but the test looks reasonable so trusting that.

File: src/dvclive/lightning.py [hunk 3]
```diff
@@ -131,6 +141,42 @@ def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None):
...
+        for t, p, _, _ in checkpoints:
+            self._logged_model_time[p] = t
+            self._all_checkpoint_paths.append(p)
+
+    def _save_checkpoints(self, checkpoint_callback: ModelCheckpoint) -> None:
+        # drop unused checkpoints
+        if not self._experiment._resume:  # noqa: SLF001
+            for p in Path(checkpoint_callback.dirpath).iterdir():
+                if str(p) not in self._all_checkpoint_paths:
+                    p.unlink(missing_ok=True)
```
Comment (daavoo) [inline src/dvclive/lightning.py:179]: Let's be clear about this in the docs

**Comment by daavoo** [comment]
I think I would like to have a `best` (or `best_only`) option for `log_model` but we can discuss separately
